crypt.go

```diff
@@ -0,0 +1,494 @@
+
+	for ; tries > 0; tries-- {
+		lastErr = nil
+
+		r := keyReader
+		keyReader = nil
+
+		passphrase, err := getPassword(sourceDevicePath, "recovery key", r)
+		if err != nil {
+			return xerrors.Errorf("cannot obtain recovery key: %w", err)
+		}
+
+		key, err := decodeRecoveryKey(passphrase)
+		if err != nil {
+			lastErr = xerrors.Errorf("cannot decode recovery key: %w", err)
+			continue
+		}
+
+		if err := activate(volumeName, sourceDevicePath, key, activateOptions); err != nil {
+			err = xerrors.Errorf("cannot activate volume: %w", err)
+			var e *exec.ExitError
+			if !xerrors.As(err, &e) {
+				return err
+			}
+			lastErr = err
+			continue
+		}
+
+		if _, err := unix.AddKey("user", fmt.Sprintf("%s:%s:reason=%d", filepath.Base(os.Args[0]), volumeName, reason), key, userKeyring); err != nil {
+			lastErr = xerrors.Errorf("cannot add recovery key to user keyring: %w", err)
+		}
+		break
+	}
+
+	return lastErr
+}
+
+func unsealKeyFromTPM(tpm *TPMConnection, k *SealedKeyObject, pin string) ([]byte, error) {
+	key, err := k.UnsealFromTPM(tpm, pin)
+	if err == ErrTPMProvisioning {
+		// ErrTPMProvisioning in this context might indicate that there isn't a valid persistent SRK. Have a go at creating one now and then
+		// retrying the unseal operation - if the previous SRK was evicted, the TPM owner hasn't changed and the storage hierarchy still
+		// has a null authorization value, then this will allow us to unseal the key without requiring any type of manual recovery. If the
+		// storage hierarchy has a non-null authorization value, ProvionTPM will fail. If the TPM owner has changed, ProvisionTPM might
+		// succeed, but UnsealFromTPM will fail with InvalidKeyFileError when retried.
+		if pErr := ProvisionTPM(tpm, ProvisionModeWithoutLockout, nil); pErr == nil {
+			key, err = k.UnsealFromTPM(tpm, pin)
+		}
+	}
+	return key, err
+}
+
+var requiresPinErr = errors.New("no PIN tries permitted when a PIN is required")
+
+type lockAccessError struct {
+	err error
+}
+
+func (e lockAccessError) Error() string {
+	return e.err.Error()
+}
+
+func (e lockAccessError) Unwrap() error {
+	return e.err
+}
+
+func isLockAccessError(err error) bool {
+	var e lockAccessError
+	return xerrors.As(err, &e)
+}
+
+func activateWithTPMKey(tpm *TPMConnection, volumeName, sourceDevicePath, keyPath string, pinReader io.Reader, pinTries int, lock bool, activateOptions []string) error {
+	var lockErr error
+	key, err := func() ([]byte, error) {
+		defer func() {
+			if !lock {
+				return
+			}
+			lockErr = LockAccessToSealedKeys(tpm)
+		}()
+
+		k, err := ReadSealedKeyObject(keyPath)
+		if err != nil {
+			return nil, xerrors.Errorf("cannot read sealed key object: %w", err)
+		}
+
+		switch {
+		case pinTries == 0 && k.AuthMode2F() == AuthModePIN:
+			return nil, requiresPinErr
+		case pinTries == 0:
+			pinTries = 1
+		}
+
+		var key []byte
+
+		for ; pinTries > 0; pinTries-- {
+			var pin string
+			if k.AuthMode2F() == AuthModePIN {
+				r := pinReader
```

Comment:
this block doesn't seem tested atm

Reply:
This requires https://github.com/snapcore/secboot/pull/34

Reply:
ok, I left some questions there last week